### PR TITLE
refactor(expr): prevent implicit array-compaction in Expression::eval

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

As title

It's still error-prone, but I can't find a better solution now.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/expr/src/expr/mod.rs [hunk 1]
```diff
@@ -56,6 +56,14 @@ pub type ExpressionRef = Arc<dyn Expression>;
 pub trait Expression: std::fmt::Debug + Sync + Send {
     fn return_type(&self) -> DataType;
 
+    fn wrapping_eval(&self, input: &DataChunk) -> Result<ArrayRef> {
```
Comment (fuyufjh) [inline src/expr/src/expr/mod.rs:59]: Please add some rustdoc

File: src/expr/src/expr/mod.rs [hunk 2]
```diff
@@ -128,7 +136,7 @@ impl RowExpression {
 
     pub fn eval(&mut self, row: &Row, data_types: &[DataType]) -> Result<ArrayRef> {
         let input = DataChunk::from_rows(slice::from_ref(row), data_types)?;
-        self.expr.eval(&input)
+        self.expr.wrapping_eval(&input)
```
Comment (fuyufjh) [inline src/expr/src/expr/mod.rs:139]: Is `eval()` useless now?

Comment (TennyZhuang) [reply src/expr/src/expr/mod.rs:139]: eval is the implementation, and wrapping_eval add some common checks.

Comment (BugenZhao) [reply src/expr/src/expr/mod.rs:139]: `wrapping` sounds like a numeric overflow wrapping to me. 😢 Name it `checked_eval` or something?

File: src/common/src/array/data_chunk_iter.rs [hunk 3]
```diff
@@ -68,6 +76,34 @@ impl<'a> Iterator for DataChunkRefIter<'a> {
...
+impl<'a> Iterator for DataChunkRefIterWithHoles<'a> {
+    type Item = Option<RowRef<'a>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let len = self.chunk.capacity();
+        let vis = self.chunk.vis();
+        if self.idx == len {
+            None
+        } else {
+            let ret = Some(if !vis.is_set(self.idx).unwrap() {
```
Comment (BugenZhao) [inline src/common/src/array/data_chunk_iter.rs:93]: Use `is_set_unchecked` here?

File: src/expr/src/expr/expr_case.rs [hunk 4]
```diff
@@ -58,40 +58,44 @@ impl Expression for CaseExpression {
     }
 
     fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
-        // TODO: we can avoid the compact here.
         let input = input.clone().compact()?;
```
Comment (BugenZhao) [inline src/expr/src/expr/expr_case.rs:61]: Still compact here?

File: src/expr/src/expr/expr_nested_construct.rs [hunk 5]
```diff
@@ -68,12 +65,15 @@ impl Expression for NestedConstructExpression {
...
-                .map_err(ExprError::Array)?;
+            )?;
+            chunk.rows_with_holes().try_for_each(|row| {
+                if let Some(row) = row {
+                    builder.append_row_ref(row)
+                } else {
+                    builder.append_null()
+                }
+            })?;
+            dbg!(&chunk);
```
Comment (BugenZhao) [inline src/expr/src/expr/expr_nested_construct.rs:76]:
```suggestion
```

File: src/expr/src/expr/mod.rs (hunk 1 above)
Comment (BugenZhao) [reply src/expr/src/expr/mod.rs:59]: +1

File: src/common/src/array/data_chunk_iter.rs (hunk 3 above)
Comment (TennyZhuang) [reply src/common/src/array/data_chunk_iter.rs:93]: NTFS, will revise vis next PR

File: src/expr/src/expr/mod.rs (hunk 2 above)
Comment (TennyZhuang) [reply src/expr/src/expr/mod.rs:139]: wrapping is just eval with a wrapper :)

Currently we do check here, but I'm thinking about something else, such as tracing span, and the check may also be updated to debug_assert! if we have some benchmarks.

File: src/expr/src/expr/mod.rs [hunk 6]
```diff
@@ -56,6 +56,14 @@ pub type ExpressionRef = Arc<dyn Expression>;
 pub trait Expression: std::fmt::Debug + Sync + Send {
     fn return_type(&self) -> DataType;
 
+    fn wrapping_eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let res = self.eval(input)?;
+
+        assert_eq!(res.len(), input.capacity());
```
Comment (xiangjinwu) [inline src/expr/src/expr/mod.rs:62]: off topic: another useful check would be that array variant matches expected return type.

**Review by xiangjinwu — APPROVED** [review]
Generally lgtm, except that `wrapping_eval` is a confusing name.

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**